src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+                    }
+                }
+            }
+        }
+
+        private static readonly ObjectPool<Queue<IOperation>> s_queuePool =
+            new ObjectPool<Queue<IOperation>>(() => new Queue<IOperation>(), 10);
+
+        private IOperation WalkDownOperationToFindParent(HashSet<IOperation> operationAlreadyProcessed, IOperation root)
+        {
+            void EnqueueChildOperations(Queue<IOperation> queue, IOperation parent)
+            {
+                // children can return null
+                foreach (var o in parent.Children.WhereNotNull())
+                {
+                    queue.Enqueue(o);
+                }
+            }
+
+            var operationQueue = s_queuePool.Allocate();
+
+            try
+            {
+                EnqueueChildOperations(operationQueue, root);
+
+                // walk down the tree to find parent operation
+                // every operation returned by the queue should already have Parent operation set
+                while (operationQueue.Count > 0)
+                {
+                    var operation = operationQueue.Dequeue();
+
+                    if (!operationAlreadyProcessed.Add(operation))
+                    {
+                        // don't process IOperation we already processed otherwise,
+                        // we can walk down same tree multiple times
+                        continue;
+                    }
+
+                    if (operation == this)
+                    {
+                        // parent found
+                        return operation.Parent;
+                    }
+
+                    // It can't filter visiting children by node span since IOperation
+                    // might have children which belong to completely different sub tree of
+                    // syntax tree
+
+                    // queue children so that we can do breadth first search
+                    EnqueueChildOperations(operationQueue, operation);
+                }
+
+                return null;
+            }
+            finally
+            {
+                operationQueue.Clear();
+                s_queuePool.Free(operationQueue);
+            }
+        }
+
+        // internal for testing
+        internal IOperation SearchParentOperation()
+        {
+            var operationAlreadyProcessed = PooledHashSet<IOperation>.GetInstance();
+
+            // start from current node since one node can have multiple operations mapped to
+            var currentCandidate = Syntax;
+
+            try
+            {
+                while (currentCandidate != null)
+                {
+                    if (!SemanticModel.Root.FullSpan.Contains(currentCandidate.FullSpan))
+                    {
+                        // reached top of parent chain
+                        break;
+                    }
+
+                    // get operation
+                    var tree = SemanticModel.GetOperationInternal(currentCandidate);
```

Comment:
GetOperation guarantee same IOperation (same instance) for same syntax node. some IOperations we synthesize which is not from BoundNode or PlaceHolder node can't be reached from syntax node.

so, even if we have 2 search happened, for same syntax node, it should return same instance of IOperation.